ActivatedPlateletSource.cpp

```diff
@@ -0,0 +1,179 @@
+    }
+
+    // Compute the source function
+    const int coarsest_ln = (coarsest_ln_in == -1 ? 0 : coarsest_ln_in);
+    const int finest_ln = (finest_ln_in == -1 ? hierarchy->getFinestLevelNumber() : finest_ln_in);
+    for (int level_num = coarsest_ln; level_num <= finest_ln; ++level_num)
+    {
+        setDataOnPatchLevel(data_idx, var, hierarchy->getPatchLevel(level_num), data_time, initial_time);
+    }
+
+    // Deallocate scratch data when needed.
+    for (const auto& local_var : vars)
+    {
+        if (!scratch_allocated[local_var])
+        {
+            auto var_db = VariableDatabase<NDIM>::getDatabase();
+            int var_scr_idx =
+                var_db->mapVariableAndContextToIndex(local_var, d_adv_diff_hier_integrator->getScratchContext());
+            d_adv_diff_hier_integrator->deallocatePatchData(var_scr_idx);
+        }
+    }
+    return;
+} // setDataOnPatchHierarchy
+
+void
+ActivatedPlateletSource::setDataOnPatch(const int data_idx,
+                                        Pointer<Variable<NDIM>> /*var*/,
+                                        Pointer<Patch<NDIM>> patch,
+                                        const double /*data_time*/,
+                                        const bool initial_time,
+                                        Pointer<PatchLevel<NDIM>> /*patch_level*/)
+{
+    Pointer<CellData<NDIM, double>> F_data = patch->getPatchData(data_idx);
+    F_data->fillAll(0.0);
+    if (initial_time) return;
+    Pointer<CartesianPatchGeometry<NDIM> > pgeom = patch->getPatchGeometry();
+    const double* const dx = pgeom->getDx();
+    // It's apparent that when Baaron wrote this, he intended I use them, so I'll ask about this (I think this relates to **)
+    Pointer<CellData<NDIM, double>> phi_a_data =
+        patch->getPatchData(d_phi_a_var, d_adv_diff_hier_integrator->getScratchContext());
+    Pointer<CellData<NDIM, double>> phi_u_data =
+        patch->getPatchData(d_phi_u_var, d_adv_diff_hier_integrator->getScratchContext());
+    Pointer<CellData<NDIM, double>> w_data =
+        patch->getPatchData(d_w_var, d_adv_diff_hier_integrator->getScratchContext());
+    const Box<NDIM>& patch_box = patch->getBox();
+    auto phi_fcn = IBAMR::getKernelAndWidth(d_kernel);
+    for (CellIterator<NDIM> ci(patch_box); ci; ci++)
+    {
+        // compute R2 R4
+        // the four patch variables we now have are phi_a and w.
+        const CellIndex<NDIM>& idx = ci();
+        // Compute source data (relaxation term)
+        // WHERE DO I GET THESE VARIABLES ** 
+        double phi_a = (*phi_a_data)(idx);
+        double phi_u = (*phi_u_data)(idx);
+        double w = (*w_data)(idx);
+        //const double eta_b = IBAMR::convolution(d_n_b_mx, phi_u, -2.0, z_data, phi_fcn.first, phi_fcn.second, idx, dx);
+        // Compute the source terms
+        const double R2 = d_a0 * phi_a * phi_a
+        const double R4 = d_a0w * (d_w_mx - w) * phi_u;
+        (*F_data)(idx) = R2 + R4; // include f^a_u?
```

Comment:
Note this should be the reaction term for the activated platelets phi_a. This only includes f^a_u. It does not include R2 or R4.

Reply:
Okay, this should be updated now